v03_pipeline/lib/tasks/reference_data/updated_reference_dataset_collection_task.py

```diff
@@ -0,0 +1,206 @@
+import os
+from datetime import datetime
+from functools import cached_property
+from types import FunctionType
+
+import hail as hl
+import luigi
+import pytz
+
+from v03_pipeline.lib.model import ReferenceDatasetCollection
+from v03_pipeline.lib.paths import valid_reference_dataset_collection_path
+from v03_pipeline.lib.reference_data.config import CONFIG
+from v03_pipeline.lib.tasks.base.base_update_task import BaseUpdateTask
+from v03_pipeline.lib.tasks.files import GCSorLocalTarget
+
+# NOTE: this is intended to replace the code in v03_pipeline/bin/write_combined_reference_ht.py, v03_pipeline/bin/write_combined_interval_reference_ht.py, and combine.py
+
+class UpdatedReferenceDatasetCollectionTask(BaseUpdateTask):
+    reference_dataset_collection = luigi.EnumParameter(enum=ReferenceDatasetCollection)
+    dataset = luigi.OptionalStrParameter()
+
+    @cached_property
+    def _destination_path(self) -> str:
+        return valid_reference_dataset_collection_path(
+            self.reference_genome,
+            self.dataset_type,
+            self.reference_dataset_collection,
+        )
+
+    def output(self) -> luigi.Target:
+        return GCSorLocalTarget(self._destination_path)
+
+    def initialize_table(self) -> hl.Table:
+        if self.dataset is None:
+            key_type = self.reference_dataset_collection.table_key_type(self.reference_genome)
+            return hl.Table.parallelize(
+                [],
+                key_type,
+                key=key_type.fields,
+                globals=hl.Struct(
+                    paths=hl.Struct(),
+                    versions=hl.Struct(),
+                    enums=hl.Struct(),
+                ),
+            )
+
+        if not hl.hadoop_exists(os.path.join(self._destination_path, '_SUCCESS')):
+            message = (
+                f'Cannot find existing valid ht at ({self._destination_path}). '
+                f'Consider omitting the --dataset flag to re-initialize the ht.'
+            )
+            raise Exception(message) # TODO: new exception type?
+
+        else:
+            return hl.read_table(self._destination_path)
+
+    def update_table(self, ht: hl.Table) -> hl.Table:
+        joined_ht = ht
```

Comment:
I bet we can get away with combining this logic now, maybe something like:

```
datasets = [self.dataset] if self.dataset is not None else self.reference_dataset_collection.datasets(self.dataset_type)

for dataset in datasets:
   dataset_ht = self._get_dataset_ht(dataset)
   // drop the field if it exists?
   // outer join
   // filter the rows where the dataset is now null
  joined_ht = self._annotate_dataset_globals(joined_ht, dataset, dataset_ht)
return joined_ht
   
```

Reply:
if that's cleaner

Reply:
I combined the logic into 1 function in the former 'combine' file!